obelisk/cpp/obelisk_cpp/include/obelisk_mujoco_sim_robot.h

```diff
@@ -0,0 +1,635 @@
+            mjv_makeScene(model_, &scn, 2000);
+            mjr_makeContext(model_, &con, mjFONTSCALE_150);
+
+            // install GLFW mouse and keyboard callbacks
+            glfwSetKeyCallback(window_, ObeliskMujocoRobot::KeyboardCallback);
+            glfwSetCursorPosCallback(window_, ObeliskMujocoRobot::MouseMoveCallback);
+            glfwSetMouseButtonCallback(window_, ObeliskMujocoRobot::MouseButtonCallback);
+            glfwSetScrollCallback(window_, ObeliskMujocoRobot::ScrollCallback);
+
+            mujoco_setup_ = true;
+            RCLCPP_INFO_STREAM(this->get_logger(), "Starting Mujoco simulation loop.");
+
+            while (!this->stop_thread_) {
+                if (!glfwWindowShouldClose(window_)) {
+                    //  Assuming MuJoCo can simulate faster than real-time, which it usually can,
+                    //  this loop will finish on time for the next frame to be rendered at 60 fps.
+                    //  Otherwise add a cpu timer and exit this loop when it is time to render.
+                    mjtNum simstart = data_->time;
+                    while (data_->time - simstart < num_steps_per_viz_ * time_step_) {
+                        {
+                            // Get the data from the shared vector safely
+                            std::lock_guard<std::mutex> lock(shared_data_mut_);
+                            for (int i = 0; i < nu_; i++) {
+                                data_->ctrl[i] = shared_data_.at(i);
+                            }
+                        }
+
+                        std::lock_guard<std::mutex> lock(sensor_data_mut_);
+                        mj_step(model_, data_);
+                    }
+
+                    // get framebuffer viewport
+                    mjrRect viewport = {0, 0, 0, 0};
+                    glfwGetFramebufferSize(window_, &viewport.width, &viewport.height);
+
+                    // update scene and render
+                    mjv_updateScene(model_, data_, &opt, NULL, &cam, mjCAT_ALL, &scn);
+                    mjr_render(viewport, &scn, &con);
+
+                    // add time stamp in upper-left corner
+                    char stamp[50];
+                    sprintf_arr(stamp, "Time = %.3f", data_->time);
+                    mjr_overlay(mjFONT_NORMAL, mjGRID_TOPLEFT, viewport, stamp, NULL, &con);
+
+                    // TODO (@zolkin): Add option to save the simulation render
+
+                    // swap OpenGL buffers (blocking call due to v-sync)
+                    glfwSwapBuffers(window_);
+
+                    // process pending GUI events, call GLFW callbacks
+                    glfwPollEvents();
+                }
+            }
+
+            // TODO: Put this here or in cleanup/shutdown?
+            // free visualization storage
+            mjv_freeScene(&scn);
+            mjr_freeContext(&con);
+
+            // free MuJoCo model and data
+            mj_deleteData(data_);
+            mj_deleteModel(model_);
```

Comment:
cleanup/shutdown. the `stop_thread_` variable should also get set in `on_cleanup` so that this loop stops in its own thread.

Reply:
Thinking about this more, I'm gonna keep these here as they are then created and destroyed all in the same place with no other functions touching them. on_cleanup and on_shutdown both end the thread.